src/lib/Repository/NameSchema/NameSchemaService.php

```diff
@@ -0,0 +1,425 @@
+        SchemaIdentifierExtractorInterface $schemaIdentifierExtractor,
+        EventDispatcherInterface $eventDispatcher,
+        array $settings = []
+    ) {
+        $this->fieldTypeRegistry = $fieldTypeRegistry;
+        // Union makes sure default settings are ignored if provided in argument
+        $this->settings = $settings + [
+                'limit' => 150,
+                'sequence' => '...',
+            ];
+        $this->eventDispatcher = $eventDispatcher;
+        $this->schemaIdentifierExtractor = $schemaIdentifierExtractor;
+    }
+
+    /**
+     * @return array<string, string> key value map of names for a language code
+     */
+    public function resolveUrlAliasSchema(Content $content, ContentType $contentType = null): array
+    {
+        $contentType ??= $content->getContentType();
+        $schemaName = $contentType->urlAliasSchema ?: $contentType->nameSchema;
+        $schemaIdentifiers = $this->schemaIdentifierExtractor->extract($schemaName);
+
+        /** @var \Ibexa\Contracts\Core\Event\ResolveUrlAliasSchemaEvent $event */
+        $event = $this->eventDispatcher->dispatch(
+            new ResolveUrlAliasSchemaEvent(
+                $schemaIdentifiers,
+                $content
+            )
+        );
+
+        $names = [];
+        $tokens = $event->getTokenValues();
+        $extractedTokens = $this->extractTokens($schemaName);
+        foreach ($tokens as $languageCode => $tokenValues) {
+            $schema = $schemaName;
+            foreach ($extractedTokens as $extractedToken) {
+                $name = $this->resolveToken($extractedToken, $tokenValues, []);
+                $schema = str_replace($extractedToken, $name, $schema);
+            }
+            $names[$languageCode] = $schema;
+        }
+
+        return $names;
+    }
+
+    /**
+     * Convenience method for resolving name schema.
+     *
+     * @param array $fieldMap
+     * @param array $languageCodes
+     *
+     * @return array
+     */
+    public function resolveNameSchema(
+        Content $content,
+        array $fieldMap = [],
+        array $languageCodes = [],
+        ContentType $contentType = null
+    ): array {
+        $contentType ??= $content->getContentType();
+
+        $languageCodes = $languageCodes ?: $content->versionInfo->languageCodes;
+
+        return $this->resolve(
+            $contentType->nameSchema,
+            $contentType,
+            $this->mergeFieldMap(
+                $content,
+                $fieldMap,
+                $languageCodes
+            ),
+            $languageCodes
+        );
+    }
+
+    /**
+     * Convenience method for resolving name schema.
+     *
+     * @param \Ibexa\Contracts\Core\Repository\Values\Content\Content $content
+     * @param array $fieldMap
+     * @param array $languageCodes
+     *
+     * @return array
+     */
+    protected function mergeFieldMap(Content $content, array $fieldMap, array $languageCodes)
```

Comment:
Can't we have a return type here? If so, the docblock should be more specific.

```suggestion
    protected function mergeFieldMap(Content $content, array $fieldMap, array $languageCodes): array
```